Introduce an optionally user-defined UUID generation and a noop system chaincode

gabre opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

The commits add two major changes:
- a noop system chaincode that is able to look up older transactions using their ID but does nothing when invoked
- an option for the user to use a custom TxID generation algorithm (specified by a hash function plus a decoding)
- support for sha256base64 (SHA256 hashing with base64 decoding)
## Description

<!--- Describe your changes in detail. -->

The noop system chaincode is loaded at system startup. It implements the following functions:
- Invoke: this part does nothing (noop), it has no internal logic. It has to be called with the function name parameter "execute" plus an argument (the payload of the transaction).
- Query: the only query implemented is "getTran" which needs a transaction ID and searches for the matching transaction.

Transactions have a new possibility: the user can select a custom algorithm for TxID generation.
- if no custom algorithm is given, more precisely if the "txIDgenerationAlg" field of ChaincodeInvocationSpec is left empty, the previously used UUID is stored as an ID
- if using CLI client "peer", one has to add "-t 'somealgorithm'" to the command where somealgorithm can be ''sha256base64"
- if using gRPC, one has to include a 'txIDgenerationAlg' string in the ChaincodeInvocationSpec (the related proto file was modified and the generated parts of source were regenerated)

A TxID generator algorithm:
- has two parts, a hash function to generate the ID and a decoding to decode user given data (arg[0] of the function that was invoked)
- hashing is done using the first argument of the invoked function provided by the user
## Motivation and Context

This modification would make the framework even more generic while not breaking any code based on previous versions of fabric:
- The user is able to create a transaction and generate an ID for it: no need to get the ID from fabric. It is important to emphasize that before this a user who created a transaction (e.g. calling invoke via gRPC) was unable to get its ID and unable to check its status etc. The modification fixes this.
## How Has This Been Tested?

Unit tests were run.
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x] Either no new documentation is required by this change, OR I added new documentation
- [x] Either no new tests are required by this change, OR I added new tests
- [x] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Gabor Hosszu gabor@digitalasset.com

gaborh-da committed:
Add noop system chaincode

gaborh-da committed:
Use payload hash instead of UUID

gaborh-da committed:
Use shim logging

gaborh-da committed:
Remove Base64 decoding of input in noop syscc

gaborh-da committed:
Implement getTran functionality for noop syscc

gaborh-da committed:
Use only user given data for txID (hashing)

gaborh-da committed:
Standardize data types in noop syscc

Input of Noop Exec is Base64 Str. This input is decoded and passed
to Noop internal logic functions as a byte array.
Fabric internally stores this input as Base64 Str. In Noop Query,
we decode this and the byte array form is returned.

gaborh-da committed:
Handle query and invoke TxIDs differently

gaborh-da committed:
Remove old utxo tests from noop directory

Old tests copied from another chaincode were removed.

gaborh-da committed:
Remove unnecessary files from directory of noop syscc

gaborh-da committed:
Move TxID generation to its old place (UUID generation)

TxID double hash generation was moved to the same place where
UUID was generated before the ID got the traditional BTC TxID
instead of UUID. There is no need for temporary IDs so they
were removed.

gaborh-da committed:
Use UUID instead of the same TxID for queries

For queries, we used TxID generated for a byte array full of zero
bytes. Now we use a really unique UUID.

gaborh-da committed:
Remove old comment leftover

gaborh-da committed:
Fix incorrect ID generation for queries

gaborh-da committed:
Add capability to set a user-given transaction ID

Transaction ID can now be specified by the user. If not specified,
UUID is used as previously.

gaborh-da committed:
Fix golint problems

gaborh-da committed:
Remove unused functions

christo4ferris commented:
I restarted the Travis build... if it fails again, more investigation is needed.

@gabre thanks for submitting this PR. It looks good on the surface, but needs some tests, please.

muralisrini commented on core/system_chaincode/noop/chaincode.go:
```diff
@@ -28,12 +28,6 @@ import (
-// components to the Bitcoin libconsensus C library for script verification. A table of UTXOs is maintained to ensure
-// each transaction is valid.
-// Documentation can be found at
-// https://github.com/hyperledger/fabric/blob/master/examples/chaincode/go/utxo/README.md
```
@gabre there are other references such as "bitcoin" (e.,g line 50) that should be removed as well.

muralisrini commented on core/system_chaincode/noop/chaincode.go:
```diff
@@ -0,0 +1,97 @@
+                        return nil, errors.New("Unsupported operation")
+	}
+
+}
```
@gabre if TXID is supplied (optionally) via "-t", querying it from Args is incorrect ? I'm not sure we need this query even ?

gabre replied:
In this case, a user given TXID (via -t) would be the transaction ID of the query itself (but queries are not stored on the ledger so it is not as useful as TXID of an 'invoke'). With this query, you are able to get the payload for a specific TXID (if you only know the TXID for any purpose).

muralisrini replied:
@gabre Ok.  I assumed this was remnant of the previous impl. where the TID was stored in Args[0]. But then you are not getting the entire payload... you are only returning Args[0]). Also, this wont work with security on as Payload is encrypted (direct Unmarshal will fail).

muralisrini commented on core/devops.go:
```diff
@@ -171,7 +171,14 @@ func (d *Devops) invokeOrQuery(ctx context.Context, chaincodeInvocationSpec *pb.
+        var userGivenID = chaincodeInvocationSpec.UserGivenID
+        var uuid string
+        if userGivenID != "" {
+                uuid = userGivenID
```
@gabre we would want to check userGivenID's uniqueness right about here I think (under a lock to be safe probably)

gabre replied:
@muralisrini I think that is not needed.That would imply enormous performance problems (e.g. if one has a big ledger it would need an expensive lookup).

muralisrini commented on protos/chaincode.proto:
```diff
@@ -94,7 +94,7 @@ message ChaincodeInvocationSpec {
     ChaincodeSpec chaincodeSpec = 1;
     //ChaincodeInput message = 2;
-
+    string userGivenID = 2;
```
seems this field is a good place for a comment explaining rationale for this work. Also perhaps change ID to TID or TransID (perhaps shorten to givenTransID or useTransID or ... )?

gabre replied:
I would vote for userGivenTxID (I think that is not too long).

corecode commented on core/devops.go:
```diff
@@ -171,7 +171,14 @@ func (d *Devops) invokeOrQuery(ctx context.Context, chaincodeInvocationSpec *pb.
 
 	// Now create the Transactions message and send to Peer.
-	uuid := util.GenerateUUID()
+        var userGivenID = chaincodeInvocationSpec.UserGivenID
```
Seems you didn't run gofmt and got indent by space instead of tab?

gabre replied:
I will correct that.

christo4ferris commented:
build failed again... needs investigation.

--- FAIL: TestSieveNoDecision (4.00s)
    obc-sieve_test.go:134: replica 0 executed 0 requests, expected 2
    obc-sieve_test.go:134: replica 1 executed 0 requests, expected 2
    obc-sieve_test.go:134: replica 2 executed 0 requests, expected 2
    obc-sieve_test.go:134: replica 3 executed 0 requests, expected 2

corecode commented:
Test timing races - it'll get better soon.

christo4ferris commented:
Ok, I have restarted again

gaborh-da committed:
Format code with gofmt

gaborh-da committed:
Add some cosmetics

gaborh-da committed:
Add a test for noop chaincode

srderson commented on core/system_chaincode/noop/chaincode.go:
```diff
@@ -0,0 +1,105 @@
+	if nil != err {
+		return nil, err
+	}
+	t.ledgerH = ledger
```
@muralisrini will this cause an issue? If the peer were to be stopped and started, is Init called again for a system chaincode or only during deployment?

srderson replied:
Confirmed with @muralisrini that this will be an issue in system chaincodes. They behave the same as regular chaincodes in the sense that anything that is not persisted in the state should be considered stateless between invocations.

On deploy `Init` will be called and set the value of `ledgerH`. However, if the peer is stopped and then started again, the protocol says the `Init` will not be called a second time because the chaincode is already deployed. This could result in an error in the `Query` function.

gabre replied:
For me, it seems that Init is called every time I start the peer. By the way I tried it with only one peer.

gabre replied:
To be honest, this is just for testing. I wanted to mock the ledger somehow. If I cannot use Init, I can also check if t.ledgerHandler is nil or not. If it is nil, I can get the ledger singleton.

ghost commented:
Can one of the admins verify this patch?

tamasblummer commented on core/system_chaincode/noop/chaincode.go:
```diff
@@ -0,0 +1,105 @@
+		logger.Infof("--> %x", args[0])
+
+		var txHashHex = args[0]
+		var tx, txerr = t.ledgerH.GetTransactionByUUID(txHashHex)
```
rename GetTransactionByUUID to GetTransactionByID

tamasblummer replied:
Better, take first 128bit of content hash to initialize UUID

gaborh-da committed:
Add possibility to use custom hash instead of UUID as TxID

This commit adds a -tid option to peer and a field to Chaincode
invocations (and replaces any previous, similar capabilities
with those) where a UUID generation algorithm can be selected.
Such algorithm contains a hash function plus a decoding. For
now, sha256 with base64 is supported. We take the first 128 bits
of the hash, and use it as a UUID.

binhn commented:
@gabre To the extent of my understanding of what you're trying to do with this pull request, may I suggest a different approach: 
1. To look up a transaction using its UUID, let's provide a GRPC API by adding the function to [api.proto](https://github.com/hyperledger/fabric/blob/master/protos/api.proto). Currently there is a REST API, but it doesn't use a GRPC API. You would prefer GRPC than REST in any case.
2. According to the [Client SDK proposal](https://github.com/hyperledger/fabric/wiki/Client-SDK-Specification), the `devops` will go away. And the SDK generates [transaction UUID](https://github.com/hyperledger/fabric/blob/master/sdk/node/src/hlc.ts) when building the transaction object. So you could leverage the same technique to implement this in your Java SDK, and you could also provide any kind of custom UUID generations.

tamasblummer commented:
@binh we are building an API in the spirit of the predicted SDK we support CLI but also extended grpc interface. Just as described in 2) the transaction ID is generated while building the transaction object on client side. The transaction ID is derived from content of the user transaction. Derivation is identical on client and server side if -tid is used.

tamasblummer commented:
Split this request into three PRs to better facilitate discussion and decision:
https://github.com/hyperledger/fabric/pull/1719
https://github.com/hyperledger/fabric/pull/1721
https://github.com/hyperledger/fabric/pull/1722
